tests/handlers/test_user_directory.py

```diff
@@ -639,6 +640,64 @@ def _add_user_to_room(
             self.hs.get_storage().persistence.persist_event(event, context)
         )
 
+    def test_making_room_public_doesnt_alter_directory_entry(self):
+        """Per-room names shouldn't go to the directory when the room becomes public.
+
+        I made this a synapse test case rather than a complement one because
+        I think this is (strictly speaking) an implementation choice. Synapse
+        has chosen to only ever use the public profile when responding to a user
+        directory search. There's no privacy leak here, because making the room
+        public discloses the per-room name.
+
+        The spec doesn't mandate anything about _how_ a user
+        should appear in a /user_directory/search result. Hypothetical example:
+        suppose Bob searches for Alice. When representing Alice in a search
+        result, it's reasonable to use any of Alice's nicknames that Bob is
+        aware of. Heck, maybe we even want to use lots of them in a combined
+        displayname like `Alice (aka "ali", "ally", "41iC3")`.
```

Comment:
There are localisation & context issues with that idea, I really think this needs spec work or being left to clients to decide. (But really, Synapse should avoid making the decision to output hardcoded English text.)

Having one canonical name — the profile name — should be understandable and clear to users about what is exposed. It also means that you don't have to fear using one single Matrix account in multiple contexts (even if the current support for having a split-life kind of account is not fully there).

Reply:
> There are localisation & context issues with that idea, I really think this needs spec work or being left to clients to decide.

Entirely agreed and not proposing this; just trying to rationalise Synapse versus Complement.